### PR TITLE
Base multiversal equality on typeclass derivation

### PR DESCRIPTION
**Pull request by odersky**
We still need to consider the idea of making equality an extension method.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

File: docs/docs/reference/other-new-features/multiversal-equality.md [hunk 1]
```diff
@@ -9,112 +9,150 @@ the fact that `==` and `!=` are implemented in terms of Java's
...
-def eqAny[L, R]: Eq[L, R] = Eq.derived
+def eqlAny[L, R]: Eql[L, R] = Eql.derived
 ```
 
-Even though `eqAny` is not declared implicit, the compiler will still
-construct an `eqAny` instance as answer to an implicit search for the
-type `Eq[L, R]`, provided that neither `L` nor `R` have `Eq` instances
-defined on them.
+Even though `eqlAny` is not declared `implied`, the compiler will still
+construct an `eqlAny` instance as answer to an implicit search for the
```
Comment (OlivierBlanvillain) [inline docs/docs/reference/other-new-features/multiversal-equality.md]: Is there also an alternative name for *implicit search*? Mixing the two terminologies here makes the sentence a bit confusing...

File: docs/docs/reference/other-new-features/multiversal-equality.md [hunk 2]
```diff
@@ -9,112 +9,150 @@ the fact that `==` and `!=` are implemented in terms of Java's
...
+ 1. `T` and `U` the same, or
+ 2. one of `T` and `U`is a subtype of the _lifted_ version of the other type, or
+ 3. neither `T` nor `U` have a _reflexive `Eql` instance_.
+
+Explanations:
+
+ - _lifting_ a type `S` means replacing all references to  abstract types
+   in covariant positions of `S` by their upper bound, and to replacing
+   all refinement types in covariant positions of `S` by their parent.
+ - a type `T` has a _reflexive `Eql` instance_ if the implicit search for `Eql[T, T]`
```
Comment (OlivierBlanvillain) [inline docs/docs/reference/other-new-features/multiversal-equality.md]: Same comment (on implicit search)

File: library/src/scala/Eq.scala [hunk 3]
```diff
@@ -7,56 +7,31 @@ import scala.collection.{GenSeq, Set}
...
-  implicit def eqInt      : Eq[Int, Int] = derived
-  implicit def eqLong     : Eq[Long, Long] = derived
-  implicit def eqFloat    : Eq[Float, Float] = derived
-  implicit def eqDouble   : Eq[Double, Double] = derived
-  implicit def eqUnit     : Eq[Unit, Unit] = derived
-
-  // true asymmetry, modeling the (somewhat problematic) nature of equals on Proxies
-  implicit def eqProxy    : Eq[Proxy, Any]     = derived
 
+  // The next three definitions can go into the companion objects of classes
```
Comment (OlivierBlanvillain) [inline library/src/scala/Eq.scala]: I agree, but then maybe then spec shouldn't mention them...

File: docs/sidebar.yml [hunk 4]
```diff
@@ -63,8 +65,6 @@ sidebar:
               url: docs/reference/contextual/relationship-implicits.html
         - title: Other New Features
           subsection:
-            - title: Multiversal Equality
-              url: docs/reference/other-new-features/multiversal-equality.html
```
Comment (OlivierBlanvillain) [inline docs/sidebar.yml]: I think this will break all links to the Multiversal Equality doc, we should keep the old url with a redirect.

**Pull request merged**